File: iso27001-springboot/src/main/java/com/iso27001/api/infrastructure/security/BruteForceGuard.java
package com.iso27001.api.infrastructure.security;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ResponseStatusException;

import java.time.Duration;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * A.9 — Brute-force lockout (5 failures → 15 min lock).
 * Redis-primary with in-process ConcurrentHashMap fallback (A.17).
 */
@Component
public class BruteForceGuard {

    private static final Logger log = LoggerFactory.getLogger(BruteForceGuard.class);
    private static final int MAX_ATTEMPTS = 5;
    private static final Duration LOCKOUT_TTL = Duration.ofMinutes(15);
    private static final String KEY_PREFIX = "bf:";

    private final StringRedisTemplate redis;

    // In-process fallback for when Redis is unavailable
    private record LocalEntry(int count, long lockedUntilMs) {}
    private final Map<String, LocalEntry> local = new ConcurrentHashMap<>();

    public BruteForceGuard(StringRedisTemplate redis) {
        this.redis = redis;
    }

    /** Throws 429 if the identifier is currently locked. */
    public void check(String identifier) {
        if (isRedisAvailable()) {
            String lockKey = KEY_PREFIX + "lock:" + identifier;
            if (Boolean.TRUE.equals(redis.hasKey(lockKey))) {
                throwLocked();
            }
        } else {
            LocalEntry entry = local.get(identifier);
            if (entry != null && entry.lockedUntilMs() > System.currentTimeMillis()) {
                throwLocked();
            }
        }
    }

    /** Increments failure counter; locks account on threshold. */
    public void recordFailure(String identifier) {
        if (isRedisAvailable()) {
            String countKey = KEY_PREFIX + "count:" + identifier;
            Long count = redis.opsForValue().increment(countKey);
            if (count != null && count == 1) {
                redis.expire(countKey, LOCKOUT_TTL);
            }
            if (count != null && count >= MAX_ATTEMPTS) {
                String lockKey = KEY_PREFIX + "lock:" + identifier;
                redis.opsForValue().set(lockKey, "1", LOCKOUT_TTL);
                log.warn("{\"event\":\"account_locked\",\"identifier_hash\":\"{}\"}", identifier.hashCode());
            }
        } else {
            local.compute(identifier, (k, entry) -> {
                int newCount = (entry == null ? 0 : entry.count()) + 1;
                long lockedUntil = newCount >= MAX_ATTEMPTS
                    ? System.currentTimeMillis() + LOCKOUT_TTL.toMillis()
                    : (entry != null ? entry.lockedUntilMs() : 0);
                return new LocalEntry(newCount, lockedUntil);
            });
        }
    }

    /** Clears failure counter on successful login. */
    public void clear(String identifier) {
        if (isRedisAvailable()) {
            redis.delete(KEY_PREFIX + "count:" + identifier);
            redis.delete(KEY_PREFIX + "lock:" + identifier);
        } else {
            local.remove(identifier);
        }
    }

    private boolean isRedisAvailable() {
        try {
            redis.getConnectionFactory().getConnection().ping();
            return true;
        } catch (Exception ignored) {
            return false;
        }
    }

    private void throwLocked() {
        throw new ResponseStatusException(
            HttpStatus.TOO_MANY_REQUESTS,
            "{\"code\":\"ACCOUNT_LOCKED\",\"message\":\"Account temporarily locked due to multiple failed login attempts\"}"
        );
    }
}

File: iso27001-nestjs/src/infrastructure/security/brute-force.guard.ts
import { Injectable, HttpException, HttpStatus } from '@nestjs/common';
import { Redis } from 'ioredis';

/**
 * A.9: Brute-force login protection.
 *
 * Tracks failed authentication attempts per account identifier (email).
 * Uses Redis when available (cross-process, survives restarts); falls back to
 * an in-process Map for dev/test environments without Redis.
 *
 * Policy:
 *   MAX_ATTEMPTS  : 5 consecutive failures trigger a lockout
 *   LOCKOUT_TTL   : 15 minutes (900 seconds)
 *   Window resets on a successful login (clear())
 */
@Injectable()
export class BruteForceGuard {
  private static readonly MAX_ATTEMPTS = 5;
  private static readonly LOCKOUT_TTL = 900; // seconds
  private static readonly KEY_PREFIX = 'brute_force:';

  // In-process fallback: { identifier -> { count, lockedUntil (epoch ms) } }
  private readonly local = new Map<string, { count: number; lockedUntil: number }>();

  constructor(private readonly redis: Redis | null) {}

  /** Throw HTTP 429 if the account is currently locked. */
  async check(identifier: string): Promise<void> {
    if (this.redis) {
      const lockedUntil = await this.redis.get(`${BruteForceGuard.KEY_PREFIX}${identifier}:locked_until`);
      if (lockedUntil && parseFloat(lockedUntil) > Date.now() / 1000) {
        this.throwLocked();
      }
    } else {
      const entry = this.local.get(identifier);
      if (entry && entry.lockedUntil > Date.now()) {
        this.throwLocked();
      }
    }
  }

  /** Increment failure counter; lock the account on threshold breach. */
  async recordFailure(identifier: string): Promise<void> {
    if (this.redis) {
      const countKey = `${BruteForceGuard.KEY_PREFIX}${identifier}:count`;
      const lockKey = `${BruteForceGuard.KEY_PREFIX}${identifier}:locked_until`;
      const count = await this.redis.incr(countKey);
      await this.redis.expire(countKey, BruteForceGuard.LOCKOUT_TTL);
      if (count >= BruteForceGuard.MAX_ATTEMPTS) {
        const lockedUntil = Math.floor(Date.now() / 1000) + BruteForceGuard.LOCKOUT_TTL;
        await this.redis.set(lockKey, lockedUntil, 'EX', BruteForceGuard.LOCKOUT_TTL);
        await this.redis.del(countKey);
      }
    } else {
      const entry = this.local.get(identifier) ?? { count: 0, lockedUntil: 0 };
      entry.count += 1;
      if (entry.count >= BruteForceGuard.MAX_ATTEMPTS) {
        entry.lockedUntil = Date.now() + BruteForceGuard.LOCKOUT_TTL * 1000;
        entry.count = 0;
      }
      this.local.set(identifier, entry);
    }
  }

  /** Clear failure counters after a successful login. */
  async clear(identifier: string): Promise<void> {
    if (this.redis) {
      await this.redis.del(
        `${BruteForceGuard.KEY_PREFIX}${identifier}:count`,
        `${BruteForceGuard.KEY_PREFIX}${identifier}:locked_until`,
      );
    } else {
      this.local.delete(identifier);
    }
  }

  private throwLocked(): never {
    throw new HttpException(
      { code: 'ACCOUNT_LOCKED', message: 'Too many failed attempts. Account temporarily locked.' },
      HttpStatus.TOO_MANY_REQUESTS,
    );
  }
}
